### PR TITLE
Use web service for tracking unit test coverage

### PR DESCRIPTION
**Pull request by ZedThree**
Adds two new Travis jobs for gathering coverage information from the unit tests and uploads to CodeCov: https://codecov.io/gh/boutproject/BOUT-dev (note: you might need to go to "branches" to see anything useful!)

One of these jobs is supposed to be "maximal", i.e. try to configure with everything we support. Unfortunately, it's not possible to use SUNDIALS with it at the mo, due to the version in Ubuntu 14. At some point, I'll make a new PR that fixes this\*.
The other job is "minimal", i.e. with as few extra features turned on as possible (except code coverage, obviously). Unfortunately, this also doesn't work as intended due to how some of the configure options interact! Another PR that fixes that will be along as well.

Also adds config for Codacy, which does static analysis online: https://app.codacy.com/app/ZedThree/BOUT-dev/dashboard
Note that that appears to be linked to my personal account -- if you sign into Codacy with your own github account, I'm not sure how the settings interact!

---
\* The plan is to download + compile it on Travis ourselves, but cache the built version so we can use it in the separate jobs.

### CODE REVIEW

File: .codacy.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+---
```
Comment (d7919) [inline .codacy.yml]: Is it worth commenting some of this to say what the different sections control?

File: .travis.yml [hunk 2]
```diff
@@ -55,6 +55,55 @@ matrix:
...
+          - libpetsc3.4.2-dev
+          - libslepc3.4.2-dev
+          - libmumps-dev
+          - libscalapack-mpi-dev
+          - libparmetis-dev
+          - libmetis-dev
+          - liblapack-dev
+          - libparpack2-dev
+    env:
+      - CONFIGURE_OPTIONS='--enable-code-coverage --enable-debug --enable-checks=3 --with-petsc --with-lapack --with-slepc --with-mumps --enable-openmp'
```
Comment (d7919) [inline .travis.yml]: I think --enable-debug should automatically set --enable-checks=3

**Comment by d7919** [comment]
This seems to also include #911 -- can we close that PR or should we remove the commit from this one?

**Comment by ZedThree** [comment]
Coverage reports were a bit inaccurate as they only included files that were "touched" by the tests. The latest commit tries to fix this by ensuring that there is a corresponding .gcda file for each .gcno file (see https://github.com/codecov/support/issues/292).

Unfortunately, we still miss stuff that doesn't get compiled in the first place, e.g. ARKODE stuff. This can be solved eventually by having enough different builds that compile with all our possible dependencies, as codecov is clever enough to merge coverage reports from multiple jobs.

**Comment by d7919** [comment]
Could we merge the two different configures for codecov into a single Travis job? This would save the ~5 minutes of environment setup time from the second case, we'd just have to make distclean in between.

**Comment by dschwoerer** [comment]
Currently I can only see that some unit tests are checked (coverage 100%).
This info is rather useless - I want to know what in src/ and include/ etc is not checked ...
Also if I click on the other 11 files, e.g. [here](https://github.com/boutproject/BOUT-dev/pull/924#issuecomment-371894191) I only see an empty page with no changes.
How can I use this?

**Comment by d7919** [comment]
If you go to the [landing page](https://codecov.io/gh/boutproject/BOUT-dev) and then click branches you can explore the coverage for the branches that currently have coverage enabled. For example [next](https://codecov.io/gh/boutproject/BOUT-dev/branch/next). At the bottom of each of these pages you get a file explorer like tool that lets you dig down by directory/file to see the coverage of different bits.

There still seems to be some issues with the accuracy/utility of the PR comments etc. but at least the branch file summary is helpful, I think.

**Comment by dschwoerer** [comment]
Yes, the branches view seems much more useful.
Thanks for the help :+1:

**Pull request merged**